lib/engine/game/base.rb

```diff
@@ -1991,35 +1993,60 @@ def ability_right_owner?(entity, ability, owner_type)
           when nil
             true
           end
-        return false unless correct_owner_type
-        return false if owner_type && (ability.owner_type.to_s != owner_type.to_s)
 
-        true
+        !!correct_owner_type
       end
 
       def ability_usable_this_or?(ability)
         !ability.count_per_or || (ability.count_this_or < ability.count_per_or)
       end
 
-      def ability_right_time?(ability, time, strict_time)
-        return false if strict_time && !ability.when
-        return true unless time
+      def ability_right_time?(ability, time, on_phase, passive_ok)
+        return true unless @round
+        return true if time == 'any' || ability.when?('any')
+        return (on_phase == ability.on_phase) || (on_phase == 'any') if on_phase
+        return false if ability.passive && !passive_ok
+        return true if ability.passive && ability.when.empty?
 
-        if (ability.type == :tile_lay) && (step = ability_blocking_step)&.is_a?(Step::SpecialTrack)
-          return step.company == ability.owner
+        # using active_step causes an infinite loop
+        current_step, current_step_name = ability_blocking_step_and_name
+
+        if (ability.type == :tile_lay) && current_step&.is_a?(Step::SpecialTrack)
+          return current_step.company == ability.owner
         end
 
-        if ability.when == 'any'
-          !strict_time
-        elsif ability.when == 'owning_corp_or_turn'
-          %w[owning_corp_or_turn track].include?(time)
-        else
-          ability.when == time.to_s
+        times = Array(time).map { |t| t == '%current_step%' ? current_step_name : t.to_s }
+        return ability.when?(*times) unless times.empty?
+
+        ability.when.any? do |ability_when|
+          case ability_when
+          when current_step_name
+            (@round.operating? && @round.current_operator == ability.corporation) ||
+              (@round.stock? && @round.current_entity == ability.player)
+          when 'owning_corp_or_turn'
+            @round.operating? && @round.current_operator == ability.corporation
+          when 'owning_player_sr_turn'
+            @round.stock? && @round.current_entity == ability.player
+          when 'other_or'
+            @round.operating? && @round.current_operator != ability.corporation
+          else
+            false
+          end
         end
       end
 
       def ability_blocking_step
-        @round.steps.find { |step| step.blocks? && !step.passed? }
+        @round.steps.find do |step|
+          # currently, abilities only care about Tracker, the is_a? check could
+          # be expanded to a list of possible classes/modules when needed
+          step.blocks? && !step.passed? && step.is_a?(Step::Tracker)
```

Comment:
why?

Reply:
do you even need ability_blocking_step?

Reply:
Yeah, it's used to sort-of figure what the active step is without calling `active_step`, since that triggers a loop of `active_step` -> what actions are available? -> what abilities are active? -> `ability_right_time?` -> `active_step`.

In 18GA, Per avoided that with these track step definitions:

* https://github.com/tobymao/18xx/pull/2947/files#diff-aa552812d2ce056cec0285385023d04fa1e6fff31b3ae39a458d2ea126bebe7e
* https://github.com/tobymao/18xx/pull/2947/files#diff-9630d124a4a448e25c7be52e26d00d79c9e1b87901e654a80c2f4effab7fb08a

Maybe something more like this could work, but company abilities that can be used any time during the corporation’s OR turn make it more challenging.